src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -617,23 +658,33 @@ internal static string GetClientFieldName(Type t, string serverDefinedName)
         /// <returns>Client PropertyInfo, or null if the method is not found.</returns>
         internal static PropertyInfo GetClientPropertyInfo(Type t, string serverDefinedName, UndeclaredPropertyBehavior undeclaredPropertyBehavior)

```

Comment:
>GetClientPropertyInfo [](http://example.com/codeflow?start=37&length=21)

Consider instead:
1) Having a Properties property (of type dictionary<string,propertyInfo>) on ODataTypeInfo.
2) The first time the Properties property is requested, generate it and then store it on the object. Since the only thread that writes to this will be the one that creates it, and it won't be set for other threads to read until it is fully populated, it does not need to be a concurrent dictionary.
3) this method just becomes:
ODataTypeInfo typeInfo = GetODataTypeInfo(t);
PropertyInfo propertyInfo = typeInfo.Properties[serverDefinedName];  
if (clientPropertyInfo == null && (undeclaredPropertyBehavior == UndeclaredPropertyBehavior.ThrowException))
    {
        throw c.Error.InvalidOperation(c.Strings.ClientType_MissingProperty(t.ToString(), serverDefinedName));
     }
 return propertyInfo;